File: FrontEnd/lex.cpp
#include "lex.h"
#include "token.h"
#include<cstdio>
#include<iostream>
Lex::Lex() {
	initTable();
	initFunc();

	curState = LexState::Start;
	bufPos = 0;
	needBack = false;
	yylineno = 0;
}

Lex::~Lex() {
}

void Lex::setFP(FILE *f) {
	fp = f;
}

void Lex::initTable() {
	// Start
	for (int i = 'a'; i <= 'z'; i++)
		table[0][i] = LexState::ID;
	for (int i = 'A'; i <= 'Z'; i++)
		table[0][i] = LexState::ID;
	for (int i = '0'; i <= '9'; i++)
		table[0][i] = LexState::digit_1;

	table[0]['i'] = LexState::int_1;
	table[0]['c'] = LexState::char_1;
	table[0]['f'] = LexState::float_1;
	table[0]['r'] = LexState::return_1;
	table[0]['b'] = LexState::bool_1;
	table[0]['w'] = LexState::while_1;
	table[0]['v'] = LexState::void_1;
	table[0]['e'] = LexState::else_1;

	table[0]['\''] = LexState::char_const_1;

	table[0][' '] = LexState::Start;

	table[0]['{'] = LexState::LCB;
	table[0]['}'] = LexState::RCB;
	table[0]['('] = LexState::LB;
	table[0][')'] = LexState::RB;
	table[0]['['] = LexState::LSB;
	table[0][']'] = LexState::RSB;
	table[0]['.'] = LexState::dot;
	table[0][';'] = LexState::semicolon;
	table[0][','] = LexState::comma;
	table[0]['='] = LexState::equal_1;

	table[0]['>'] = LexState::greater_1;
	table[0]['<'] = LexState::less_1;

	table[0]['&'] = LexState::logic_and_1;
	table[0]['|'] = LexState::logic_or_1;

	table[0]['*'] = LexState::multiply;
	table[0]['/'] = LexState::d_1;
	//table[0]['/'] = LexState::divide;
	table[0]['+'] = LexState::add;
	table[0]['-'] = LexState::substract;

	// key
	for (int j = (int)LexState::int_1; j < (int)LexState::else_4; j++) {
		for (int i = 'a'; i <= 'z'; i++)
			table[j][i] = LexState::ID;
		for (int i = 'A'; i <= 'Z'; i++)
			table[j][i] = LexState::ID;
		for (int i = '0'; i <= '9'; i++)
			table[j][i] = LexState::ID;
		for (int i = ' '; i <= '/'; i++)
			table[j][i] = LexState::identifier;
		for (int i = ':'; i <= '@'; i++)
			table[j][i] = LexState::identifier;
		for (int i = '['; i <= '`'; i++)
			table[j][i] = LexState::identifier;
		for (int i = '{'; i <= '~'; i++)
			table[j][i] = LexState::identifier;
	}

	// int_k

	_setK(LexState::int_1, "nt", LexState::int_k);

	// float_k

	_setK(LexState::float_1, "loat", LexState::float_k);

	// char_k
	_setK(LexState::char_1, "har", LexState::char_k);
	// bool_k
	_setK(LexState::bool_1, "ool", LexState::bool_k);

	_setK(LexState::return_1, "eturn", LexState::return_k);

	_setK(LexState::while_1, "hile", LexState::while_k);

	_setK(LexState::void_1, "oid", LexState::void_k);

	_setK(LexState::else_1, "lse", LexState::else_k);

	table[(int)LexState::int_1]['f'] = LexState::if_2;
	_setKTI(LexState::if_2, LexState::if_k);
	table[(int)LexState::float_1]['o'] = LexState::for_2;
	table[(int)LexState::for_2]['r'] = LexState::for_3;
	_setKTI(LexState::for_3, LexState::for_k);

	//
	_setKTI(LexState::ID, LexState::identifier);
	_meetLetter(LexState::ID, LexState::ID);
	_meetDigit(LexState::ID, LexState::ID);
	//

	_meetDigit(LexState::digit_1, LexState::digit_1);
	_meetDigit(LexState::digit_2, LexState::digit_2);
	_meetLetter(LexState::digit_1, LexState::error);
	_meetLetter(LexState::digit_2, LexState::error);
	_setKTI(LexState::digit_1, LexState::int_const);
	table[(int)LexState::digit_1]['.'] = LexState::digit_2;
	_setKTI(LexState::digit_2, LexState::float_const);

	// =
	
	for (int i = 0; i < 127; i++) {
		table[(int)LexState::equal_1][i] = LexState::assign;
		table[(int)LexState::less_1][i] = LexState::less;
		table[(int)LexState::greater_1][i] = LexState::greater;
	}
		
	table[(int)LexState::equal_1]['='] = LexState::double_equal;
	table[(int)LexState::less_1]['='] = LexState::less_equal;
	table[(int)LexState::greater_1]['='] = LexState::greater_equal;

	table[(int)LexState::logic_and_1]['&'] = LexState::logic_and;
	table[(int)LexState::logic_or_1]['|'] = LexState::logic_or;

	//
	
	//table[(int)LexState::d_2]['\n'] = LexState::Start;
	for (int i = 0; i < 127; i++) {
		table[(int)LexState::d_1][i] = LexState::divide;
		table[(int)LexState::d_2][i] = LexState::d_2;
		table[(int)LexState::d_4][i] = LexState::d_3;
		table[(int)LexState::d_3][i] = LexState::d_3;
	}
	table[(int)LexState::d_1]['/'] = LexState::d_2;
	table[(int)LexState::d_1]['*'] = LexState::d_3;
	table[(int)LexState::d_3]['*'] = LexState::d_4;
	table[(int)LexState::d_4]['/'] = LexState::Start;
}

void Lex::initFunc() {
	void(*f)() = [] {};
	for (int i = 0; i < 50; i++)
		func[i] = f;

	//
	//func[Token::int_const] = [] {printf("INTC"); };
	func[Token::LCB] = [] {};
	func[Token::RCB] = [] {  };
	func[Token::add] = [] {lexVal = new SSNode(SSType::add, yylineno, yytext); };
	func[Token::substract] = [] {lexVal = new SSNode(SSType::sub, yylineno, yytext); };
	func[Token::multiply] = [] {lexVal = new SSNode(SSType::mult, yylineno, yytext); };
	func[Token::divide] = [] {lexVal = new SSNode(SSType::div, yylineno, yytext); };
	//func[Token::su] = [] {};
	func[Token::identifier] = [] {lexVal = new SSNode(SSType::identifier, yylineno, yytext); };

	func[Token::int_k] = [] {lexVal = new SSNode(SSType::int_k, yylineno, yytext); };
	func[Token::float_k] = [] {lexVal = new SSNode(SSType::float_k, yylineno, yytext); };
	func[Token::void_k] = [] {lexVal = new SSNode(SSType::void_k, yylineno, yytext); };

	func[Token::else_k] = [] { /*lexVal = new SSNode(SSType::else_k, yylineno, yytext);*/ };
	func[Token::int_const] = [] {lexVal = new SSNode(SSType::int_const, yylineno, yytext); };
	func[Token::float_const] = [] {lexVal = new SSNode(SSType::float_const, yylineno, yytext); };

}

void Lex::_meetDigit(LexState s, LexState f) {
	for (int i = '0'; i <= '9'; i++)
		table[(int)s][i] = f;
}
void Lex::_meetLetter(LexState s, LexState f) {
	for (int i = 'a'; i <= 'z'; i++)
		table[(int)s][i] = f;
	for (int i = 'A'; i <= 'Z'; i++)
		table[(int)s][i] = f;
}
void Lex::_setKTI(LexState s, LexState f) {
	for (int i = ' '; i <= '/'; i++)
		table[(int)s][i] = f;
	for (int i = ':'; i <= '@'; i++)
		table[(int)s][i] = f;
	for (int i = '['; i <= '`'; i++)
		table[(int)s][i] = f;
	for (int i = '{'; i <= '~'; i++)
		table[(int)s][i] = f;
}
void Lex::_setK(LexState k_1, std::string s, LexState k) {
	int len = s.length();
	for (int i = 0; i < len; i++) {
		table[(int)k_1 + i][s[i]] = (LexState)((int)k_1 + i + 1);
	}
	_setKTI((LexState)((int)k_1 + len), (LexState)(int)k);
}
Token Lex::lex() {

	char c;
	curState = LexState::Start;
	bufPos = 0;
	//int fff = 0;
	while ((c = getC()) != EOF) {
		//std::cout << fff++;
		if (c == '\n') {
			yylineno++;
			if (curState == LexState::d_2) {
				curState = LexState::Start;
			}
			continue;
		}
		if (c == '\t')
			c = ' ';
		LexState res = table[(int)curState][c];
		if ((int)res >= 100 && res < LexState::error) {
			if ((int)res < 120 || res >= LexState::greater || res == LexState::divide) {
				needBack = true;
				charBack = c;
				yytext[bufPos] = '\0';
			}
			else {
				yytext[bufPos] = c;
				yytext[bufPos + 1] = '\0';
			}
			
			//printf("%s\n", yytext);

			Token t = (Token)((int)res - 100);
			token_list.push_back(t);
			//std::cout << (int)t;
			func[t]();

			return t;
		}
		else {
			if (c != ' ') {
				yytext[bufPos] = c;
				bufPos++;
			}
			
			curState = res;
		}
		
	}
	return Token::END;
}
int Lex::getC() {
	if (needBack) {
		needBack = false;
		return charBack;
	}
	else
		return fgetc(fp);
}

void Lex::print() {
	int id_index = 1, int_const_index = 1, float_const_index = 1, char_const_index = 1;

	for (int i = 0; i < token_list.size(); i++) {
		Token t = token_list[i];
		if (t == Token::identifier) {
			std::cout << "<" << "id, " << id_index++ << ">";
		}
		else if (t == Token::int_const) {
			std::cout << "<" << "int_const, " << int_const_index++ << ">";
		}
		else if (t == Token::float_const) {
			std::cout << "<" << "float_const, " << float_const_index++ << ">";
		}
		else if (t >= Token::int_k && t <= Token::else_k) {
			std::cout << "<" << "k, " << (t - Token::int_k + 1) << ">";
		}
		else if (t >= Token::LCB && t <= Token::assign) {
			std::cout << "<" << "p, " << (t - Token::LCB + 1) << ">";
		}
		else if (t == Token::char_const) {
			std::cout << "<" << "char_const, " << char_const_index++ << ">";
		}
		//std::cout << std::endl;
	}
}

void Lex::print_for_json() {
	int id_index = 0, int_const_index = 0, float_const_index = 0, char_const_index = 0;
	std::cout << "\"Token\":" << "\"";
	for (int i = 0; i < token_list.size(); i++) {
		Token t = token_list[i];
		if (t == Token::identifier) {
			std::cout << "<" << "id, " << id_index++ << ">";
		}
		else if (t == Token::int_const) {
			std::cout << "<" << "int_const, " << int_const_index++ << ">";
		}
		else if (t == Token::float_const) {
			std::cout << "<" << "float_const, " << float_const_index++ << ">";
		}
		else if (t >= Token::int_k && t <= Token::else_k) {
			std::cout << "<" << "k, " << (t - Token::int_k + 1) << ">";
		}
		else if (t >= Token::LCB && t <= Token::assign) {
			std::cout << "<" << "p, " << (t - Token::LCB + 1) << ">";
		}
		else if (t == Token::char_const) {
			std::cout << "<" << "char_const, " << char_const_index++ << ">";
		}
	}
	std::cout << "\",";
}

File: c-compiler-win32-x64/resources/app/lex.js
const LexState = {
    Start : 0, 
	ID : 1, 
	digit_1 : 2, digit_2 : 3, 
	int_1 : 4, int_2 : 4 , int_3 : 4,
	if_2 : 7,
	char_1 : 8, char_2 : 4, char_3 : 4, char_4 : 4,
	float_1 : 12, float_2 : 4, float_3 : 4, float_4 : 4, float_5 : 4,
	for_2 : 17, for_3 : 4,
	void_1 : 19, void_2 : 4, void_3 : 4, void_4 : 4,
	return_1 : 23, return_2 : 4, return_3 : 4, return_4 : 4, return_5 : 4, return_6 : 4,
	while_1 : 29, while_2 : 4, while_3 : 4, while_4 : 4, while_5 : 4,
	bool_1 : 34, bool_2 : 4, bool_3 : 4, bool_4 : 4,
	else_1 : 38, else_2 : 4, else_3 : 4, else_4 : 4,
	char_const_1 : 42, char_const_2 : 4, char_const_3 : 4,
	equal_1 : 44,
	greater_1 : 4,
	less_1 : 4,
	logic_or_1 : 4,
	logic_and_1 : 4,

	// fin
	int_const : 100, float_const : 4, char_const : 4, 
	int_k : 4, float_k : 4, char_k : 4, bool_k : 4, void_k : 4,
	return_k : 4, for_k : 4, while_k : 4, if_k : 4, else_k : 4,
	identifier : 4,
	
	//120 { } ( ) . ; , 
	LCB : 120, RCB : 4, LB : 4, RB : 4, LSB : 4, RSB : 4, dot : 4, semicolon : 4, comma : 4,  
	// * / + -
	multiply : 129, divide : 4, add : 4, substract : 4,
	logic_or : 4, logic_and : 4,
	// . >: <: > < :
	double_equal : 135, unequal : 4,greater_equal : 4, less_equal : 4, greater : 4, less : 4, assign : 4,

	end : 200
	

};
let n = 0;
for( s in LexState){
	//console.log(n);
	LexState[s] = n++;
	
}
Object.freeze(LexState);

class Lex{

	constructor(df){
        this.curState = LexState.Start;
        this.table = [];
        this.funcTable = [];
		this.defaultFunc = df;
		this.buf = "";
		this.initTable();
		this.initFunc(df);
	}
	initTable(){
		for(let i = LexState.Start; i <= LexState.assign; i++){
			this.table[i] = [];
		}

		this._meetDigit(LexState.Start, LexState.digit_1);
		this._meetLetter(LexState.Start, LexState.ID);
		this.table[LexState.Start]['i'] = LexState.int_1;
		this.table[0]['c'] = LexState.char_1;
		this.table[0]['f'] = LexState.float_1;
		this.table[0]['r'] = LexState.return_1;
		this.table[0]['b'] = LexState.bool_1;
		this.table[0]['w'] = LexState.while_1;
		this.table[0]['v'] = LexState.void_1;
		this.table[0]['e'] = LexState.else_1;

		this.table[0]['\''] = LexState.char_const_1;

		this.table[0][' '] = LexState.Start;

		this.table[0]['{'] = LexState.LCB;
		this.table[0]['}'] = LexState.RCB;
		this.table[0]['('] = LexState.LB;
		this.table[0][')'] = LexState.RB;
		this.table[0]['['] = LexState.LSB;
		this.table[0][']'] = LexState.RSB;
		this.table[0]['.'] = LexState.dot;
		this.table[0][';'] = LexState.semicolon;
		this.table[0][','] = LexState.comma;
		this.table[0]['='] = LexState.equal_1;

		this.table[0]['>'] = LexState.greater_1;
		this.table[0]['<'] = LexState.less_1;

		this.table[0]['&'] = LexState.logic_and_1;
		this.table[0]['|'] = LexState.logic_or_1;

		this.table[0]['*'] = LexState.multiply;
		this.table[0]['/'] = LexState.divide;
		this.table[0]['+'] = LexState.add;
		this.table[0]['-'] = LexState.substract;


		// key
		for(let j = LexState.int_1; j <= LexState.else_4; j++){
			for(let i = 'a'.charCodeAt(); i <= 'z'.charCodeAt(); i++){
				this.table[j][String.fromCharCode(i)] = LexState.ID;
			}
			for(let i = 'A'.charCodeAt(); i <= 'Z'.charCodeAt(); i++){
				this.table[j][String.fromCharCode(i)] = LexState.ID;
			}
			for(let i = '0'.charCodeAt(); i <= '9'.charCodeAt(); i++){
				this.table[j][String.fromCharCode(i)] = LexState.ID;
			}
			for(let i = ' '.charCodeAt(); i <= '/'.charCodeAt(); i++){
				this.table[j][String.fromCharCode(i)] = LexState.identifier;
			}
			for(let i = ':'.charCodeAt(); i <= '@'.charCodeAt(); i++){
				this.table[j][String.fromCharCode(i)] = LexState.identifier;
			}
			for(let i = '['.charCodeAt(); i <= '`'.charCodeAt(); i++){
				this.table[j][String.fromCharCode(i)] = LexState.identifier;
			}
			for(let i = '{'.charCodeAt(); i <= '~'.charCodeAt(); i++){
				this.table[j][String.fromCharCode(i)] = LexState.identifier;
			}
		}

		// int_k

		this._setK(LexState.int_1, "nt", LexState.int_k);

		// float_k

		this._setK(LexState.float_1, "loat", LexState.float_k);

		// char_k
		this._setK(LexState.char_1, "har", LexState.char_k);
		// bool_k
		this._setK(LexState.bool_1, "ool", LexState.bool_k);

		this._setK(LexState.return_1, "eturn", LexState.return_k);

		this._setK(LexState.while_1, "hile", LexState.while_k);

		this._setK(LexState.void_1, "oid", LexState.void_k);

		this._setK(LexState.else_1, "lse", LexState.else_k);

		this.table[LexState.int_1]['f'] = LexState.if_2;
		this._setKTI(LexState.if_2, LexState.if_k);
		this.table[LexState.float_1]['o'] = LexState.for_2;
		this.table[LexState.for_2]['r'] = LexState.for_3;
		this._setKTI(LexState.for_3, LexState.for_k);

		//
		this._setKTI(LexState.ID, LexState.identifier);
		this._meetLetter(LexState.ID, LexState.ID);
		this._meetDigit(LexState.ID, LexState.ID);
		//

		this._meetDigit(LexState.digit_1, LexState.digit_1);
		this._meetDigit(LexState.digit_2, LexState.digit_2);
		this._meetLetter(LexState.digit_1, LexState.error);
		this._meetLetter(LexState.digit_2, LexState.error);
		this._setKTI(LexState.digit_1, LexState.int_const);
		this.table[LexState.digit_1]['.'] = LexState.digit_2;
		this._setKTI(LexState.digit_2, LexState.float_const);

		// char_const
		for (let i = 0; i < 127; i++) {
			this.table[LexState.char_const_1][String.fromCharCode(i)] = LexState.char_const_3;
			this.table[LexState.char_const_2][String.fromCharCode(i)] = LexState.char_const_3;
		}
		this.table[LexState.char_const_1]['\\'] = LexState.char_const_2;
		this.table[LexState.char_const_3]['\''] = LexState.char_const;
		// =
		
		for (let i = ' '.charCodeAt(); i < '~'.charCodeAt(); i++) {
			this.table[LexState.equal_1][String.fromCharCode(i)] = LexState.assign;
			this.table[LexState.less_1][String.fromCharCode(i)] = LexState.less;
			this.table[LexState.greater_1][String.fromCharCode(i)] = LexState.greater;
		}
		
		this.table[LexState.equal_1]['='] = LexState.double_equal;
		this.table[LexState.less_1]['='] = LexState.less_equal;
		this.table[LexState.greater_1]['='] = LexState.greater_equal;

		this.table[LexState.logic_and_1]['&'] = LexState.logic_and;
		this.table[LexState.logic_or_1]['|'] = LexState.logic_or;

	}
	initFunc(df){
		for(let i = LexState.int_const; i <= LexState.assign; i++){
			this.table[i] = df;
		}
	}
	setDeaultFunc(func){
		/*for(i of LexState){
			
		}*/
	}
	setFunc(s, func){
		this.funcTable[s] = func;
	}
	_meetDigit(s, f){
        for(let i = '0'.charCodeAt(); i <= '9'.charCodeAt(); i++){
            this.table[s][String.fromCharCode(i)] = f;
        }
	}
	_meetLetter(s, f){
        for (let i = 'a'.charCodeAt(); i <= 'z'.charCodeAt(); i++)
			this.table[s][String.fromCharCode(i)] = f;
	    for (let i = 'A'.charCodeAt(); i <= 'Z'.charCodeAt(); i++)
			this.table[s][String.fromCharCode(i)] = f;
	}
	_setKTI(s, f){
        for (let i = ' '.charCodeAt(); i <= '/'.charCodeAt(); i++)
			this.table[s][String.fromCharCode(i)] = f;
        for (let i = ':'.charCodeAt(); i <= '@'.charCodeAt(); i++)
			this.table[s][String.fromCharCode(i)] = f;
        for (let i = '['.charCodeAt(); i <= '`'.charCodeAt(); i++)
			this.table[s][String.fromCharCode(i)] = f;
        for (let i = '{'.charCodeAt(); i <= '~'.charCodeAt(); i++)
			this.table[s][String.fromCharCode(i)] = f;
            
	}
	_setK(k_1, str, k){
        let len = str.length;
        for(let i = 0; i < len; i++){
            this.table[k_1 + i][str[i]] = k_1 + i + 1;
        }
        this._setKTI(k_1 + len, k);
	}
	reset(){
		this.buf = "";
		this.curState = LexState.Start;
	}
	lex(ch){
		/*if(this.table[this.curState] === undefined){
			console.log(this.curState);
		}*/
		if(ch === '\t')ch = ' ';
		/*console.log(this.curState);
		console.log("+");
		console.log(ch);*/
		//console.log("!!!!!!!!!!!!! " + ch + " !!!!!!!");
		let res = this.table[this.curState][ch];
		if(res >= LexState.int_const && res <= LexState.end){
			if ((res < LexState.LCB && res !== LexState.char_const) || res >= LexState.greater) {
				
				this.curState = LexState.Start;
				this.defaultFunc(this.buf, res);
				//console.log("-----------------------------------");
				//console.log(this.buf + " " + ch);
				this.buf = "";

				this.lex(ch);
			}
			else {
				this.buf += ch;
				
				this.curState = LexState.Start;
				this.defaultFunc(this.buf, res);
				//console.log(this.buf + " " + ch);
				this.buf = "";
				
			}
			/*console.log("this.curState");
			console.log(this.curState);*/
			/*let f = this.funcTable[res];
			if(f != null){
				f(res);
			}*/
			//console.log(this.buf);
			
		}else{
			/*if(ch != ' '){
				this.buf += ch;
			}*/
			this.buf += ch;
			this.curState = res;
		}
		/*console.log("=");
		console.log(this.curState);
		console.log("-------------------------------------------");*/

	}
}



exports.LexState = LexState;
exports.Lex = Lex;